File: testtillpressure_np.c
/*
 * This is a simple program to test the Tillotson EOS library.
 *
 * Author:   Christian Reinhardt
 * Date:     04.03.2019
 * Modified: 05.03.2019
 *
 * Calculate the pressure of a material with tillPressureSoundNP() and check if it agrees with
 * tillPressure().
 */
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "tillotson.h"

#define INDEX(i, j) ((i*tillmat->nTableV) + (j))

void main(int argc, char **argv) {
    // Tillotson EOS library
	TILLMATERIAL *tillmat;
    double dKpcUnit = 2.06701e-13;
    double dMsolUnit = 4.80438e-08;
    double vmax = 100.0;
    int iMat = GRANITE;
    // Define the size of the grid
    double rhomin = 1e-4;
	double rhomax = 10.0;
	double umax = 25.0;
	double umin = 0.0;
    double T_min = 0.0;
    double T_max = 1e4;
	int nRho = 1000;
	int nU = 1000;
	int nV = 1000;
	int nT = 100;
    double drho, dU, dT;
	double rho, u, T;
	FILE *fp = NULL;
	int i, j;

	fprintf(stderr, "Initializing material...\n");
	tillmat = tillInitMaterial(iMat, dKpcUnit, dMsolUnit);
	tillInitLookup(tillmat, nRho, nV, rhomin, rhomax, vmax);
    fprintf(stderr, "Done.\n");

    /*
     * Zoom in to the expanded states.
     */
    rhomax = 1.1*tillmat->rho0;
    umax = 1.1*tillmat->us2;

    fprintf(stderr, "rhomin= %g rhomax= %g umin= %g umax= %g\n", rhomin, rhomax, umin, umax);

    drho = (rhomax-rhomin)/(nRho-1);
    dU = (umax-umin)/(nU-1);
    dT = (T_max-T_min)/(nT-1);

	/*
	 * Print P on a rho x u grid.
	 */	
	fp = fopen("testtillpressure_np.txt", "w");
	assert(fp != NULL);

	fprintf(stderr, "Printing grid using tillPressureSoundNP()...\n");

	/* Print a rho x u grid. */
	for (i=0; i<nU; i+=1)
	{
		u = umin + i*dU;
		for (j=0; j<nRho; j+=1)
		{
		    rho = rhomin + j*drho;
			fprintf(fp," %15.7E", tillPressureSoundNP(tillmat, rho, u, NULL));
		}
		fprintf(fp,"\n");
	}
	fclose(fp);

	fp = fopen("testtillpressure.txt", "w");
	assert(fp != NULL);

	fprintf(stderr, "Printing grid using tillPressure()...\n");

	/* Print a rho x u grid. */
	for (i=0; i<nU; i+=1)
	{
        u = umin + i*dU;

		for (j=0; j<nRho; j+=1)
		{
            rho = rhomin + j*drho;

			fprintf(fp," %15.7E", tillPressure(tillmat, rho, u));
		}
		fprintf(fp,"\n");
	}
	fclose(fp);

	fp = fopen("testtillpressure_np_diff.txt", "w");
	assert(fp != NULL);

	/* print a rho x u grid. */
	for (i=0; i<nU; i+=1)
	{
        u = umin + i*dU;

		for (j=0; j<nRho; j+=1)
		{
            rho = rhomin + j*drho;

            double P;
            P = fabs(tillPressureSoundNP(tillmat, rho, u, NULL)-tillPressure(tillmat, rho, u));

            if (P > 0.0) {
                fprintf(fp," %2i", 0);
            } else {
                fprintf(fp," %2i", 1);
            }
#if 0
            fprintf(fp," %15.7E", tillPressureSoundNP(tillmat, rho, u, NULL)-
                    tillPressure(tillmat, rho, u));
#endif
		}
		fprintf(fp,"\n");
	}
	fclose(fp);

    /* Plot where P is negative. */
	fp = fopen("testtillpressure_np_region.txt", "w");
	assert(fp != NULL);

	/* print a rho x u grid. */
	for (i=0; i<nU; i+=1)
	{
        u = umin + i*dU;

		for (j=0; j<nRho; j+=1)
		{
            rho = rhomin + j*drho;
            if (tillPressureSoundNP(tillmat, rho, u, NULL) <= 0.0)
//            if (tillPressureSound(tillmat, rho, u, NULL) <= 0.0)
            {
                fprintf(fp," %2i", 0);
            } else {
                fprintf(fp," %2i", 1);
            }
		}
		fprintf(fp,"\n");
	}
	fclose(fp);

    /*
     * Plot a few isotherms.
     */
	fp = fopen("testtillpressure_np_rhot.txt", "w");
	assert(fp != NULL);

	for (i=0; i<nRho; i+=1)
	{
        rho = rhomin + i*drho;

		fprintf(fp,"%15.7E", rho);

		for (j=0; j<nT; j+=1)
		{
            T = T_min + j*dT;
            u = eosURhoTemp(tillmat, rho, T);
			fprintf(fp," %15.7E", tillPressureSoundNP(tillmat, rho, u, NULL));
		}
		fprintf(fp,"\n");
	}
	fclose(fp);

	tillFinalizeMaterial(tillmat);
}

File: testtillpressure_np.py
#!/usr/bin/env python2
"""
This script plots the results of plottillpressure.c.
"""
from matplotlib import *
rcParams['font.family'] = 'serif'
from numpy import *
from matplotlib.pyplot import *

"""
Setup the plot.
"""
# Set a font
rcParams['font.family'] = 'serif'
rcParams['font.size'] = 10.0

# Legend
# mpl.rcParams['legend.handlelength']  = 2.9
rcParams['legend.handlelength']  = 0.5
rcParams['legend.frameon']       = False
rcParams['legend.numpoints']     = 1
rcParams['legend.scatterpoints'] = 1

# Adjust axes line width
rcParams['axes.linewidth']   = 0.5

# Adjust ticks
rcParams['xtick.major.size'] = 4
rcParams['xtick.minor.size'] = 2
rcParams['ytick.major.size'] = 4
rcParams['ytick.minor.size'] = 2

# Adjust Font Size
rcParams['xtick.labelsize']  = 'xx-small'
rcParams['ytick.labelsize']  = 'xx-small'
rcParams['axes.labelsize']   = 'x-small'

# Set Up Figure, Single Column MNRAS
fig = gcf()
ax = gca()
fig, ax = subplots(1,1)
fig.set_size_inches(8.27*0.39,8.27*(6./8.)*0.39)

# Physical constants used for unit convertion (cgs)
KBOLTZ = 1.38e-16		# bolzman constant in cgs
MHYDR  = 1.67e-24		# mass of hydrogen atom in grams
MSOLG  = 1.99e33		# solar mass in grams
GCGS   = 6.67e-8		# G in cgs
KPCCM  = 3.085678e21		# kiloparsec in centimeters

# These two values define the unit system we use
dKpcUnit  = 2.06701e-13	        # kiloparsec in code units
dMsolUnit = 4.80438e-08	        # solar mass in code units

# Mass of Earth
MEarth = 5.98e27		# g
MSun   = 1.989e33	        # g

"""
Convert kboltz/mhydrogen to system units, assuming that
G == 1.
"""
# code energy per unit mass --> erg per g
dErgPerGmUnit = GCGS*dMsolUnit*MSOLG/(dKpcUnit*KPCCM)
# code density --> g per cc
dGmPerCcUnit = (dMsolUnit*MSOLG)/pow(dKpcUnit*KPCCM,3.0)
# code time --> seconds
dSecUnit = sqrt(1/(dGmPerCcUnit*GCGS))

# The units are not nescessary but can be derived from the quanities above
Lunit = dKpcUnit*KPCCM
Munit = dMsolUnit*MSOLG

"""
data1 = loadtxt('out.old.dat',skiprows=3)
data2 = loadtxt('out.dat',skiprows=3)
"""

"""
# Load values for rho and u
data = loadtxt('testtillpressure_np_diff.txt')

rho = data[:, 0]
u   = data[:, 1]
T   = data[:, 2]

print "rho_min=", min(rho), "rho_max=", max(rho)
print "u_min  =", min(u), "u_max  =", max(u)

# Convert to cgs
rho /= dGmPerCcUnit
u   /= dErgPerGmUnit

#data /= (dErgPerGmUnit*dGmPerCcUnit)
#data /= (dGmPerCcUnit*dErgPerGmUnit)

rho_min = min(rho)
rho_max = max(rho)

u_min   = min(u)
u_max   = max(u)

T_min   = min(T)
T_max   = max(T)

print "rho_min=", min(rho), "rho_max=", max(rho)
print "u_min  =", min(u), "u_max  =", max(u)
"""
# Reference density (material dependent)
rho0 = 2.7
us   = 3.5e10
us2  = 1.8e11

rho_min = 1e-4
rho_max = 10.0
u_min   = 0.0
u_max   = 25.0

rho_min = 1e-4
rho_max = 8.0602
u_min   = 0.0
u_max   = 19.8035

# Convert to cgs
rho_min *= dGmPerCcUnit
rho_max *= dGmPerCcUnit

u_min   *= dErgPerGmUnit
u_max   *= dErgPerGmUnit

print dGmPerCcUnit
print dErgPerGmUnit

print "rho_min=", rho_min, "rho_max=", rho_max
print "u_min  =", u_min, "u_max  =", u_max

"""
Plot the difference between tillPressureSoundNP and tillPressure.
"""
data = loadtxt('testtillpressure_np_diff.txt')

print where(fabs(data) < 1e-10)

imshow(data, origin='lower', extent=(rho_min, rho_max, u_min, u_max), aspect='auto')

plot([rho0, rho0], [u_min, u_max], '--', color='red')
plot([rho_min, rho_max], [us, us], '--', color='red')
plot([rho_min, rho_max], [us2, us2], '--', color='red')

xlabel("Density [code units]")
ylabel("Int. energy [code units]")

colorbar()

savefig('testtillpressure_np_diff.png', dpi=300, bbox_inches='tight')

fig.clear()

"""
Plot where the pressure is negative.
"""
data = loadtxt('testtillpressure_np_region.txt')

imshow(data, origin='lower', extent=(rho_min, rho_max, u_min, u_max), aspect='auto')

plot([rho0, rho0], [u_min, u_max], '--', color='red')
plot([rho_min, rho_max], [us, us], '--', color='red')
plot([rho_min, rho_max], [us2, us2], '--', color='red')

xlabel("Density [code units]")
ylabel("Int. energy [code units]")

colorbar()

savefig('testtillpressure_np_region.png', dpi=300, bbox_inches='tight')

fig.clear()

"""
Plot P(rho, T).
"""
data = loadtxt('testtillpressure_np_rhot.txt')

rho = data[:,0]

for i in range(1,size(data[1,:]),1):
    plot(data[:,0],data[:,i],'-',color='red',markersize=1,label='T')

rho_min = min(rho)
rho_max = max(rho)

xlim(rho_min, rho_max)
#ylim(0, 4)

xlabel("Density [code units]")
ylabel("Pressure [code units]")

savefig('testtillpressure_np_rhot.png', dpi=300, bbox_inches='tight')

show()


